File: backtracking/subsets/subsets.go
package subsets

func subsets(nums []int) [][]int {
	list := []int{}
	ans := [][]int{}

	traverse(&list, &ans, 0, nums)

	return ans
}

func traverse(list *[]int, ans *[][]int, start int, nums []int) {
	temp := make([]int, len(*list))
	copy(temp, *list)
	*ans = append(*ans, temp)

	for i := start; i < len(nums); i++ {
		*list = append(*list, nums[i])
		traverse(list, ans, i+1, nums)
		*list = (*list)[:len(*list)-1]
	}

}

File: backtracking/subsets/subsets.java
class Solution {
  public List<List<Integer>> subsets(int[] nums) {
    List<Integer> list = new ArrayList<>();
    List<List<Integer>> ans = new ArrayList<>();

    traverse(0, list, ans, nums);

    return ans;
  }

  private void traverse(int start, List<Integer> list, List<List<Integer>> ans, int[] nums) {
    ans.add(new ArrayList<>(list));

    for (int i = start; i < nums.length; i++) {
      list.add(nums[i]);
      traverse(i + 1, list, ans, nums);
      list.remove(list.size() - 1);
    }
  }
}
